// @vueuse/motion/dist/nuxt.cjs
'use strict';

const path = require('path');
const url = require('url');
const kit = require('@nuxt/kit');

const DEFAULTS = {};
const CONFIG_KEY = "motion";
const module$1 = kit.defineNuxtModule({
  meta: {
    name: "@vueuse/motion",
    configKey: CONFIG_KEY,
    compatibility: {
      nuxt: ">=3.0.0",
      bridge: true
    }
  },
  defaults: DEFAULTS,
  setup(options, nuxt) {
    const templatesDir = url.fileURLToPath(new URL("./templates", (typeof document === 'undefined' ? new (require('u' + 'rl').URL)('file:' + __filename).href : (document.currentScript && document.currentScript.src || new URL('nuxt.cjs', document.baseURI).href))).toString());
    kit.addTemplate({
      fileName: "motion.config.js",
      src: path.resolve(templatesDir, "motion.config.js")
    });
    kit.addPluginTemplate({
      src: path.resolve(templatesDir, "motion.js"),
      fileName: "motion.js",
      options
    });
    if (!nuxt.options.build.transpile)
      nuxt.options.build.transpile = [];
    const transpileList = ["defu", "@vueuse/motion", "@vueuse/shared", "@vueuse/core"];
    transpileList.forEach((pkgName) => {
      if (!nuxt.options.build.transpile.includes(pkgName))
        nuxt.options.build.transpile.push(pkgName);
    });
    if (!nuxt.options.alias)
      nuxt.options.alias = {};
    if (!nuxt.options.alias.tslib)
      nuxt.options.alias.tslib = "tslib/tslib.es6.js";
  }
});

module.exports = module$1;
